scripts/appbuilder

```diff
@@ -212,34 +212,28 @@ get_image()
 # export a flattened copy of the container
 export_image()
 {
-    TEMP_FILE=$(mktemp)
-    TEMP_DIR=$(mktemp -d)
     APPENV_FILENAME="appenv.json"
-
-    # note that we have to start up the image to flatten it
-    TEMP_INAME=$(docker run -d $IMAGE_NAME)
-    TEMP_APPENV=$APPENV_FILENAME-$TEMP_INAME
+    TEMP_APPENV=$APPENV_FILENAME-$(date +%s)
     
     docker image inspect --format='{{json .Config}}' $IMAGE_NAME > $TEMP_APPENV
+    LOWER_DIR=$(docker image inspect --format='{{json .GraphDriver.Data.LowerDir}}' $IMAGE_NAME | tr -d '"')
+    UPPER_DIR=$(docker image inspect --format='{{json .GraphDriver.Data.UpperDir}}' $IMAGE_NAME | tr -d '"')
+    WORK_DIR=$(docker image inspect --format='{{json .GraphDriver.Data.WorkDir}}' $IMAGE_NAME | tr -d '"')
 
-    docker stop $TEMP_INAME >/dev/null
-    docker export $TEMP_INAME -o $TEMP_FILE
-    docker rm $TEMP_INAME >/dev/null
-    if [ -z $DELETE_IMAGE ]; then
-        docker rmi -f $IMAGE_NAME
-    fi
-
-    # create the 'appdir' from the docker export
-    tar xf $TEMP_FILE -C $TEMP_DIR
-    mv $TEMP_APPENV $TEMP_DIR/$APPENV_FILENAME
-    rm -rf $OUTPUT_NAME $TEMP_FILE
+    # cleanup $OUTPUT_NAME if it exists
+    rm -rf $OUTPUT_NAME
 
+    MNT_PNT_DIR=$(mktemp -d)
+    sudo mount -t overlay overlay -olowerdir=$LOWER_DIR,upperdir=$UPPER_DIR,workdir=$WORK_DIR $MNT_PNT_DIR
+    # change ownership back to current effective user and group
+    sudo chown -R $USER:$(id -ng) $MNT_PNT_DIR
     if [ "$OUTPUT_FORMAT" == "dir" ]; then
-        mv $TEMP_DIR $OUTPUT_NAME
+        cp -r $MNT_PNT_DIR $OUTPUT_NAME
     elif [ "$OUTPUT_FORMAT" == "cpio" ]; then
-        myst mkcpio $TEMP_DIR $OUTPUT_NAME
+        myst mkcpio $MNT_PNT_DIR $OUTPUT_NAME
     fi
-    rm -rf $TEMP_DIR
+    sudo umount $MNT_PNT_DIR
```

Comment:
if the script can fail between the mount and unmount it will be left around. There is a way of having a cleanup action when the script shuts down even where there is an early termination and can allow for cleanup if needed. Please consider this as having mounts not getting cleaned up can be annoying.

Reply:
Added a cleanup flow based on bash's `trap` mechanism.